spec/datadog/kit/events_spec.rb

```diff
@@ -10,72 +10,95 @@
       end
+      trace = trace_op.flush!
+      expect(trace.send(:meta)).to include('usr.id' => '42', 'appsec.events.users.login.success.foo' => 'bar')
     end
+  end
 
-    context 'login failure' do
-      it 'sets event tracking key on trace' do
-        trace_op.measure('root') do
-          described_class.track_login(trace_op, user: { id: '42' }, success: false)
-        end
-        trace = trace_op.flush!
-        expect(trace.send(:meta)).to include('appsec.events.users.login.failure.track' => 'true')
+  describe '#track_login_failure' do
+    it 'sets event tracking key on trace' do
+      trace_op.measure('root') do
+        described_class.track_login_failure(trace_op, user_id: '42', user_exists: true)
       end
+      trace = trace_op.flush!
+      expect(trace.send(:meta)).to include('appsec.events.users.login.failure.track' => 'true')
+    end
 
-      it 'sets user id on trace' do
-        trace_op.measure('root') do
-          described_class.track_login(trace_op, user: { id: '42' }, success: false)
-        end
-        trace = trace_op.flush!
-        expect(trace.send(:meta)).to include('appsec.events.users.login.failure.usr.id' => '42')
+    it 'sets user id on trace' do
+      trace_op.measure('root') do
+        described_class.track_login_failure(trace_op, user_id: '42', user_exists: true)
       end
+      trace = trace_op.flush!
+      expect(trace.send(:meta)).to include('appsec.events.users.login.failure.usr.id' => '42')
+    end
 
-      it 'sets other keys on trace' do
-        trace_op.measure('root') do
-          described_class.track_login(trace_op, user: { id: '42' }, success: false, foo: 'bar')
-        end
-        trace = trace_op.flush!
-        expect(trace.send(:meta)).to include('appsec.events.users.login.failure.foo' => 'bar')
+    it 'sets user existence on trace' do
+      trace_op.measure('root') do
+        described_class.track_login_failure(trace_op, user_id: '42', user_exists: true)
       end
+      trace = trace_op.flush!
+      expect(trace.send(:meta)).to include('appsec.events.users.login.failure.usr.exists' => 'true')
+    end
+
+    it 'sets user non-existence  on trace' do
+      trace_op.measure('root') do
+        described_class.track_login_failure(trace_op, user_id: '42', user_exists: false)
+      end
+      trace = trace_op.flush!
+      expect(trace.send(:meta)).to include('appsec.events.users.login.failure.usr.exists' => 'false')
+    end
+
+    it 'sets other keys on trace' do
+      trace_op.measure('root') do
+        described_class.track_login_failure(trace_op, user_id: '42', user_exists: true, foo: 'bar')
+      end
+      trace = trace_op.flush!
+      expect(trace.send(:meta)).to include('appsec.events.users.login.failure.foo' => 'bar')
     end
   end
 
   describe '#track' do
+    it 'rejects unexpected namespaces' do
+      trace_op.measure('root') do
+        expect { described_class.track(:foo, 'bar', trace_op) }.to raise_error ArgumentError
```

Comment:
Minor: While I don't think it's a good idea to match on the exact string, I suggest also not being too vague to avoid a test passing due to other kinds of argument errors.

Aka something like `.to raise_error(ArgumentError, /namespace cannot be/)`